File: 프로그래머스/레벨2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.js
function solution(queue1, queue2) {
    // 빠른 반환 : 합이 홀수
    let q1_sum = queue1.reduce((a, b) => BigInt(a) + BigInt(b), BigInt(0));
    let total_sum = q1_sum + queue2.reduce((a, b) => BigInt(a) + BigInt(b), BigInt(0));
    if (total_sum & BigInt(1)) return -1;

    // 종료 조건 : 현재 합이 전체 합의 절반
    let current_sum = q1_sum;
    let target_sum = total_sum >> BigInt(1);
    
    // 원형 큐 : 두 큐를 결합
    let n = queue1.length;
    let size = 2 * n;
    let q = queue1.concat(queue2);

    // 투 포인터 : i나 j가 한바퀴를 돌면 종료
    let cnt = 0, i = 0, j = n - 1;
    while (i < size && j < size) {
        if (current_sum === target_sum) return cnt;

        if (current_sum > target_sum)   current_sum -= BigInt(q[i++ % size]);
        else                            current_sum += BigInt(q[++j % size]);

        cnt++;
    }

    return -1;
}

File: 프로그래머스/레벨2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.cpp
#include <vector>
#include <numeric>

using namespace std;

int solution(vector<int> queue1, vector<int> queue2) {
    // 빠른 반환 : 합이 홀수    
    long long q1_sum = accumulate(queue1.begin(), queue1.end(), 0LL);
    long long total_sum = q1_sum + accumulate(queue2.begin(), queue2.end(), 0LL);
    if (total_sum & 1) return -1;

    // 종료 조건 : 현재 합이 전체 합의 절반
    long long current_sum = q1_sum;
    long long target_sum = total_sum >> 1;
        
    // 원형 큐 : 두 벡터를 결합    
    int n = queue1.size();
    int size = 2 * n;
    long long* q = new long long[size];
    copy(queue1.begin(), queue1.end(), q);
    copy(queue2.begin(), queue2.end(), q + n);

    // 투 포인터 : i나 j가 한바퀴를 돌면 종료
    int cnt = 0, i = 0, j = n - 1;
    while (i < size && j < size) {
        if (current_sum == target_sum) return cnt;

        if (current_sum > target_sum)   current_sum -= q[i++ % size];
        else                            current_sum += q[++j % size];

        cnt++;
    }

    return -1;
}